### PR TITLE
Add API and address config to enable testing bootz service in tests

### PR DESCRIPTION
**Pull request by mojiiba**
This pull addresses https://github.com/openconfig/bootz/issues/20. The unit tests will be completed before merge, but before doing  that just want to make sure you agree with the way i am doing it.

### CODE REVIEW

**Comment by gmacf** [comment]
I'm not a huge fan of putting test-only code in the Bootz serving code. Is there a way to instead change the public API to return the information we need to do tests?

**Comment by mojiiba** [comment]
> I'm not a huge fan of putting test-only code in the Bootz serving code. Is there a way to instead change the public API to return the information we need to do tests?

Yes, I agree with you and the pr does that. The test in this pr is just a unit test for the APIs.

**Comment by gmacf** [comment]
> > I'm not a huge fan of putting test-only code in the Bootz serving code. Is there a way to instead change the public API to return the information we need to do tests?
> 
> Yes, I agree with you and the pr does that. The test in this pr is just a unit test for the APIs.

Sorry, what I mean is I don't think we should be adding APIs to the Bootz server that will only be used for testing.

For example, you have added a method to the service that will only ever be used by tests"IsChassisConnected". Is there a way we can test if the chassis is connected by inspecting the response or error of the GetBootstrapData gRPC? Similar applies to getting boot status and resetting the server. IMO these are all things that should be evaluated by looking at the public API, not through a test-only one.

**Comment by mojiiba** [comment]
no test has no way to access those info since the test is a separate entity and have no access to the message and logs transferred between the bootzserver and  the router, unless by adding APIs. We can also
(1) create a separate server along tests that only uses entity manager from the bootz repo and include all testing apis.  However, there will lead to duplicate code and impose a risk that test developers may introduce a behaviors that does not comply with the reference implementation. 
(2) may use grpc proxy between the router and bootzserevr to inspect the messages between the two, but this will make the testing complicated and probably flaky. Also, not sure how dhcp flow can be validated in this approach.

We can have an offline discussion on this if you want. 

@marcushines  what do u suggest?

**Comment by gmacf** [comment]
> no, a test has no way to access those info since the test is a separate entity and have no access to the message and logs transferred between the bootzserver and the router, unless by adding APIs. We can also (1) create a separate server along tests that only uses entity manager from the bootz repo and include all testing apis. However, there will lead to duplicate code and impose a risk that test developers may introduce a behaviors that does not comply with the reference implementation. (2) may use grpc proxy between the router and bootzserevr to inspect the messages between the two, but this will make the testing complicated and probably flaky. Also, not sure how dhcp flow can be validated in this approach.
> 
> We can have an offline discussion on this if you want.
> 
> @marcushines what do u suggest?

Thanks for clarifying. I'll set up another meeting offline for us to discuss this.

File: server/service/service_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package service
```
Comment (melzhan) [inline server/service/service_test.go:1]: Please add a license header to all new files. I've just submitted a PR that adds them to all the other files so just look at one of them for reference.

Comment (mojiiba) [reply server/service/service_test.go:1]: sure, I will

File: client/client.go [hunk 2]
```diff
@@ -55,6 +55,16 @@ func pemEncodeCert(contents string) string {
 	return strings.Join([]string{"-----BEGIN CERTIFICATE-----", contents, "-----END CERTIFICATE-----"}, "\n")
 }
 
+// convert address to localhost when no ip is specefied
```
Comment (melzhan) [inline client/client.go:58]: Comments that are complete sentences should be capitalized and punctuated like standard English sentences.

typo: specified

File: server/server.go [hunk 3]
```diff
@@ -25,10 +25,20 @@ import (
 )
 
 var (
-	port              = flag.String("port", "", "The port to start the Bootz server on localhost")
+	bootzAddress      = flag.String("address", "8008", "The [ip:]port to listen for the bootz request. when ip is not given, the server will listen on localhost.")
 	artifactDirectory = flag.String("artifact_dir", "../testdata/", "The relative directory to look into for certificates, private keys and OVs.")
 )
 
+// convert address to localhost when no ip is specefied
```
Comment (melzhan) [inline server/server.go:32]: as above, typo and make full sentence

File: server/service/service_test.go (hunk 1 above)
Comment (melzhan) [reply server/service/service_test.go:1]: Also note that packages must be preceded with a one line comment explaining the purpose of the package

**Comment by mojiiba** [comment]
will open separate pulls for test apis

**Pull request closed without merging**